tests/src/Functional/AccessTest.php

```diff
@@ -0,0 +1,251 @@
+
+/**
+ * Tests access to collabora routes.
+ */
+class AccessTest extends BrowserTestBase {
+
+    use MediaTypeCreationTrait;
+    use TestFileCreationTrait;
+
+    /**
+     * {@inheritdoc}
+     */
+    protected static $modules = [
+        'node',
+        'media',
+        'collabora_online',
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    protected $defaultTheme = 'stark';
+
+    /**
+     * Tests access to core paths, to verify that the tests work correctly.
+     */
+    public function testCorePathsAccess(): void {
+        $this->createMediaType('file', ['id' => 'document']);
+        $media = $this->createMediaEntity('document');
+        $media_id = $media->id();
+
+        $authenticated = $this->createUser();
+        $paths = [
+            // Test the front page as an example that everybody can access.
+            '/',
+            // Test an administration page that only admin can see.
+            '/admin/config',
+            // Test the user route.
+            '/user/' . $authenticated->id(),
+            // Test the core media route for reference.
+            '/media/' . $media->id() . '/edit',
+        ];
+        $users = [
+            'anonymous' => new AnonymousUserSession(),
+            'authenticated' => $authenticated,
+            'admin' => $this->createUser(admin: TRUE),
+        ];
+
+        // Build a report and assert the full result at once.
+        // This provides a very complete picture in case the assertion fails.
+        $this->assertSameReport(
+            [
+                '/' => ['anonymous', 'authenticated', 'admin'],
+                '/admin/config' => ['admin'],
+                '/user/' . $authenticated->id() => ['authenticated', 'admin'],
+                "/media/$media_id/edit" => ['admin'],
+            ],
+            $this->buildAccessReport($paths, $users),
+            'Users with access to given paths',
+        );
+    }
+
+    /**
+     * Tests a scenario when only the administrator has access.
+     */
+    public function testOnlyAdminHasAccess(): void {
+        $this->createMediaType('file', ['id' => 'document']);
+        $media = $this->createMediaEntity('document');
+
+        $paths = [
+            '/cool/view/' . $media->id(),
+            '/cool/edit/' . $media->id(),
+        ];
+        $users = [
+            'anonymous' => new AnonymousUserSession(),
+            'authenticated' => $this->createUser(),
+            'admin' => $this->createUser(admin: TRUE),
+        ];
+
+        // Both routes are only accessible for admin.
+        $this->assertSameReport(
+            [
+                '/cool/view/1' => ['admin'],
+                '/cool/edit/1' => ['admin'],
+            ],
+            $this->buildAccessReport($paths, $users),
+            'Users with access to given paths',
+        );
+    }
+
+    /**
+     * Tests a scenario where specific permissions are given to users.
+     */
+    public function testCollaboraMediaPermissions(): void {
```

Comment:
you could rather build a simpler test that receives data from a yeilding function (doesn't need to be marked as a provider, we can just loop)

Reply:
@drishu I am happy with the new version of the test, are you?
I really like the big assertion that gives an overview of expected vs actual access, as opposed to single asserts where a failing test only gives us one piece of information at a time.

Reply:
(Francesco cannot click "resolve", so I do.)